### DISCUSSION TITLE
AssertionError: Im2Col_0: ConvolutionInputGenerator1D works only for 1D convs

### DISCUSSION THREAD

**Discussion by sakshigoel155** [Xilinx/finn#602] [Q&A] [answered]
Hi,

I have a model with 4 2D convolution layers followed by 3 dense layers. I generate an ONNX model for it with input size (1,1,4,512).
While trying out DataflowBuildConfig on it to generate a FINN representation, I get the following error.

Final outputs will be generated in output_estimates_only
Build log is at output_estimates_only/build_dataflow.log
Running step: step_qonnx_to_finn [1/8]
Running step: step_tidy_up [2/8]
Running step: step_streamline [3/8]

Traceback (most recent call last):
  File "/workspace/finn/src/finn/builder/build_dataflow.py", line 166, in build_dataflow_cfg
    model = transform_step(model, cfg)
  File "/workspace/finn/src/finn/builder/build_dataflow_steps.py", line 295, in step_convert_to_hls
    model = model.transform(to_hls.InferConvInpGen())
  File "/workspace/finn-base/src/finn/core/modelwrapper.py", line 141, in transform
    (transformed_model, model_was_changed) = transformation.apply(
  File "/workspace/finn/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py", line 201, in apply
    assert is_1d_convolution, (
AssertionError: Im2Col_0: ConvolutionInputGenerator1D works only for 1D convs

=======================================================================
I also tried the flow using bnn-pynq tutorial and after the streamlining step, while converting to HLS with createdataflowpartition, I get the same error, 
![error](https://user-images.githubusercontent.com/74159729/170526086-974ecf0f-d0ec-42e2-9d68-1e06e1244d8c.png)

I looked at the error further and it seems, it can only work for square input images. My input is not square input image but a 4X512 input, which uses EEG time series data from 4 channels, 512 samples at a time. 

The graph after streamlining, just before 
parent_model = model.transform(CreateDataflowPartition()), without fpgadataflow.convert_to_hls_layers.InferConvInpGen, looks like in the attached image
![QAT-1_IR1 onnx](https://user-images.githubusercontent.com/74159729/170522042-00d654fe-69af-421a-9420-9afa77331a4d.png)

Can someone please help in finding out a solution on how can i go ahead?. Will be really thankful if someone can help me, as my project deadline is near. 
Thanks

**Comment by joshua99911** [comment]
Hi, running into almost the exact same error and didn't want to open a new thread if this one exists.

Did you by chance find a way to fix the error or does anyone else know how?

**Reply by o-ibrahim33** [reply]
Same error here, any one has an answer?

**Reply by fpjentzsch** [reply, marked as answer]
Hi, this case of a non-square feature map should be supported by the RTL version of the ConvolutionInputGenerator node (as opposed to the HLS implementation that is the default). You could try to enable it by either
a) settting `force_rtl_conv_inp_gen = True` in the build cfg (see [here](https://github.com/Xilinx/finn/blob/b3bdff118ae076cb776af6e51ddc28eeaa0d6390/src/finn/builder/build_dataflow_config.py#L284))
b) inferring the layer using `InferConvInpGen(use_rtl_variant=True)`

I would also recommend to use the FINN dev branch, if you are not doing that already.

**Reply by o-ibrahim33** [reply]
@fpjentzsch it worked! Thank you very much for your help
